File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/Schedule.java
package net.nemerosa.ontrack.job;

import lombok.Data;
import org.apache.commons.lang3.StringUtils;

import java.util.concurrent.TimeUnit;

@Data
public class Schedule {

    private final long initialPeriod;
    private final long period;
    private final TimeUnit unit;

    public String getPeriodText() {
        if (period <= 0) {
            return "Manually";
        } else if (period == 1) {
            return "Every " + StringUtils.substringBeforeLast(unit.name().toLowerCase(), "s");
        } else {
            return "Every " + period + " " + unit.name().toLowerCase();
        }
    }

    public long toMiliseconds() {
        return TimeUnit.MILLISECONDS.convert(period, unit);
    }

    public static Schedule everySeconds(long seconds) {
        return new Schedule(0, seconds, TimeUnit.SECONDS);
    }

    public static Schedule everyMinutes(long minutes) {
        return new Schedule(0, minutes, TimeUnit.MINUTES);
    }

    public static final Schedule NONE = everySeconds(0);
    public static final Schedule EVERY_SECOND = everySeconds(1);
    public static final Schedule EVERY_MINUTE = everyMinutes(1);
    public static final Schedule EVERY_DAY = new Schedule(0, 1, TimeUnit.DAYS);

    public Schedule after(int initial) {
        return new Schedule(
                initial,
                period,
                unit
        );
    }

    public boolean sameDelayThan(Schedule schedule) {
        return this.period == schedule.period && this.unit.equals(schedule.unit);
    }

    public Schedule convertTo(TimeUnit target) {
        return new Schedule(
                target.convert(initialPeriod, unit),
                target.convert(period, unit),
                target
        );
    }
}

File: ontrack-job/src/main/java/net/nemerosa/ontrack/job/Schedule.kt
package net.nemerosa.ontrack.job

import org.apache.commons.lang3.StringUtils
import java.util.concurrent.TimeUnit

data class Schedule(
        val initialPeriod: Long = 0,
        val period: Long,
        val unit: TimeUnit,
        val cron: String? = null,
) {
    val periodText: String
        get() = when {
            cron != null && cron.isNotBlank() -> cron
            period <= 0 -> "Manually"
            period == 1L -> "Every " + StringUtils.substringBeforeLast(unit.name.lowercase(), "s")
            else -> "Every " + period + " " + unit.name.lowercase()
        }

    fun toMiliseconds(): Long = TimeUnit.MILLISECONDS.convert(period, unit)

    fun after(initial: Int): Schedule {
        check(cron == null || cron.isBlank()) { "Setting an initial delay is not supported for cron schedules." }
        return Schedule(
                initial.toLong(),
                period,
                unit
        )
    }

    companion object {

        fun cron(expression: String) = Schedule(
            initialPeriod = 0,
            period = 0,
            unit = TimeUnit.MILLISECONDS,
            cron = expression,
        )

        @JvmStatic
        fun everySeconds(seconds: Long): Schedule {
            return Schedule(0, seconds, TimeUnit.SECONDS)
        }

        @JvmStatic
        fun everyMinutes(minutes: Long): Schedule {
            return Schedule(0, minutes, TimeUnit.MINUTES)
        }

        @JvmField
        val NONE = everySeconds(0)
        @JvmField
        val EVERY_SECOND = everySeconds(1)
        @JvmField
        val EVERY_MINUTE = everyMinutes(1)
        @Suppress("unused")
        @JvmField
        val EVERY_HOUR = Schedule(0, 1, TimeUnit.HOURS)
        @JvmField
        val EVERY_DAY = Schedule(0, 1, TimeUnit.DAYS)
        @JvmField
        val EVERY_WEEK = Schedule(0, 7, TimeUnit.DAYS)
    }
}
